File: golang/ex11.go
package main

import (
	"fmt"
)

type Empregado interface {
	CalcularSalario() float64
}

type EmpregadoHorista struct {
	HorasTrabalhadas float64
	ValorHora        float64
}

func (e EmpregadoHorista) CalcularSalario() float64 {
	return e.HorasTrabalhadas * e.ValorHora
}

type EmpregadoAssalariado struct {
	SalarioMensal float64
}

func (e EmpregadoAssalariado) CalcularSalario() float64 {
	return e.SalarioMensal
}

func ExibirSalario(emp Empregado) {
	fmt.Printf("Salário: %.2f\n", emp.CalcularSalario())
}

func main() {
	empregado1 := EmpregadoHorista{HorasTrabalhadas: 40, ValorHora: 15.0}
	empregado2 := EmpregadoAssalariado{SalarioMensal: 3000.0}

	ExibirSalario(empregado1)
	ExibirSalario(empregado2)
}

File: python/ex11.py
from abc import ABC, abstractmethod


class Empregado(ABC):
    @abstractmethod
    def calcular_remuneracao(self):
        pass


class EmpregadoHorista(Empregado):
    def __init__(self, nome_empregado, horas_trabalhadas, valor_por_hora):
        self.nome_empregado = nome_empregado
        self.horas_trabalhadas = horas_trabalhadas
        self.valor_por_hora = valor_por_hora

    def calcular_remuneracao(self):
        return self.horas_trabalhadas * self.valor_por_hora


class EmpregadoAssalariado(Empregado):
    def __init__(self, nome_empregado, salario_mensal):
        self.nome_empregado = nome_empregado
        self.salario_mensal = salario_mensal

    def calcular_remuneracao(self):
        return self.salario_mensal


def executar():
    empregado_horista = EmpregadoHorista("Carlos", 40, 15)  # 40 horas trabalhadas a R$15/h
    empregado_assalariado = EmpregadoAssalariado("Ana", 3000)  # Salário fixo de R$3000

    print(f"Remuneração do {empregado_horista.nome_empregado}: R${empregado_horista.calcular_remuneracao():.2f}")
    print(f"Remuneração da {empregado_assalariado.nome_empregado}: R${empregado_assalariado.calcular_remuneracao():.2f}")


if __name__ == "__main__":
    executar()
